sklearn/_fast_gradient_boosting/binning.py

```diff
@@ -0,0 +1,140 @@
+from ..utils.validation import check_is_fitted
+from ._binning import _map_to_bins
+from .types import X_DTYPE, X_BINNED_DTYPE
+
+
+def _find_binning_thresholds(data, max_bins=256, subsample=int(2e5),
+                             random_state=None):
+    """Extract feature-wise quantiles from numerical data.
+
+    Return
+    ------
+    binning_thresholds: tuple of arrays
+        For each feature, stores the increasing numeric values that can
+        be used to separate the bins. len(binning_thresholds) == n_features.
+    """
+    if not (2 <= max_bins <= 256):
+        raise ValueError('max_bins={} should be no smaller than 2 '
+                         'and no larger than 256.'.format(max_bins))
+    rng = check_random_state(random_state)
+    if subsample is not None and data.shape[0] > subsample:
+        subset = rng.choice(np.arange(data.shape[0]), subsample)
+        data = data[subset]
+
+    percentiles = np.linspace(0, 100, num=max_bins + 1)
+    end = percentiles.shape[0]  # no negative indexing!
+    percentiles = percentiles[1:end - 1]
+    binning_thresholds = []
+    for f_idx in range(data.shape[1]):
+        col_data = np.ascontiguousarray(data[:, f_idx], dtype=X_DTYPE)
+        distinct_values = np.unique(col_data)
+        if len(distinct_values) <= max_bins:
+            end = distinct_values.shape[0]  # no negative indexing!
+            midpoints = (distinct_values[:end - 1] + distinct_values[1:])
+            midpoints *= .5
+        else:
+            # We sort again the data in this case. We could compute
+            # approximate midpoint percentiles using the output of
+            # np.unique(col_data, return_counts) instead but this is more
+            # work and the performance benefit will be limited because we
+            # work on a fixed-size subsample of the full data.
+            midpoints = np.percentile(col_data, percentiles,
+                                      interpolation='midpoint').astype(X_DTYPE)
+        binning_thresholds.append(midpoints)
+    return binning_thresholds
+
+
+class _BinMapper(BaseEstimator, TransformerMixin):
+    """Transformer that maps a dataset into integer-valued bins.
+
+    The bins are created in a feature-wise fashion, using quantiles so that
+    each bins contains approximately the same number of samples.
+
+    Large datasets are subsampled, but the feature-wise quantiles should
+    remain stable.
+
+    If the number of unique values for a given feature is less than
+    ``max_bins``, then the unique values of this feature are used instead of
+    the quantiles.
+
+    Parameters
+    ----------
+    max_bins : int, optional (default=256)
+        The maximum number of bins to use. If for a given feature the number of
+        unique values is less than ``max_bins``, then those unique values
+        will be used to compute the bin thresholds, instead of the quantiles.
+    subsample : int or None, optional (default=1e5)
```

Comment:
somehow you have different default values here compared to the function up there. Not sure why.

Reply:
Good catch, lightgbm is using 2e5 (param name is `bin_construct_sample_cnt`) and I forgot to update it here.